src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/BuildPropertyPage/DefineConstantsCSharpValueProvider.cs

```diff
@@ -7,25 +7,28 @@ namespace Microsoft.VisualStudio.ProjectSystem.Properties;
 
 [ExportInterceptingPropertyValueProvider(ConfiguredBrowseObject.DefineConstantsProperty, ExportInterceptingPropertyValueProviderFile.ProjectFile)]
 [AppliesTo(ProjectCapability.CSharpOrFSharp)]
-internal class DefineConstantsValueProvider : InterceptingPropertyValueProviderBase
+internal class DefineConstantsCSharpValueProvider : InterceptingPropertyValueProviderBase
 {
+    private const string DefineConstantsRecursivePrefix = "$(DefineConstants)";
+
     private readonly IProjectAccessor _projectAccessor;
     private readonly ConfiguredProject _project;
-
-    internal const string DefineConstantsRecursivePrefix = "$(DefineConstants)";
-
+    private HashSet<string> _removedValues = [];
```

Comment:
I'm not sure it's safe to store this state here. These providers are generally safe to call concurrently. By tracking this state here, that's no longer the case. It might work most of the time, but it makes me a little nervous. I'll spend some more time here to see if I can come up with any suggestions.

Reply:
I've removed this state and instead handle collapsing multiple values down into one inside the CPS control, as provided by a regex rule in editor metadata. Thanks, your comment prompted me to rethink how this worked.